Move "Display Rotation" and stretching options to display layout editor.

LunaMoo opened the pull request:
Seems fitting since it's not really a performance setting where it is now, also makes one option less in graphics setting menu.
 Also fix portrait rotation representation in the editor with auto setting and minor cleanup.

 Edit: this kind of goes along the recent cleaning spirit:P, but if the main change is not ok let me know so I can revert it and just leave the minor fix and cleaning.

LunaMoo committed:
Move "Display Rotation" to display layout editor.
Fix portrait rotation representation in the editor with auto setting.
Minor cleanup.

daniel229 commented:
Can you add "Display Rotation" to the windows menu bar under the "Emulation".

LunaMoo commented:
Ok, not sure if that's the right place, but I guess it can be treated same as umd switch also something we can't physicaly do(well rotating TV for example;p).

LunaMoo committed:
Add Display Rotation to windows menu bar.

CarlKenner commented:
Sounds like a bad idea to me. It's not really a performance setting, but it is really a game specific graphics hack.

daniel229 commented:
@LunaMoo Thanks.

LunaMoo commented:
@CarlKenner it's not really a graphics hack since it doesn't affect game rendering, it's more of a feature that on desktop can't be physically done, so just like UMD switch which already is under emulation settings it seems to fit there. Either way that was only about creating new option in windows menu bar and that's divided differently from main menus on simply emulation, debug and game settings.

 Being or not being performance settings I mentioned in context of where the option is placed now in ppsspp GUI, since along #8171 we want to clear as much clutter as we can from the main menus graphics setting is really tragic with that. The display layout editor screen seems much better place for it since it's already centered around moving and resizing display. It will still be in the graphics section, just one screen deeper.

@daniel229 you're welcome, through it will not depend on me if that get's there or not;3.

hrydgard commented:
This has some odd issues: 

   * The rotation picker drifts upwards and disappears of the screen as you resize the window
   * The screen overlaps the blue borders when switching to landscape


![capture](https://cloud.githubusercontent.com/assets/130929/11127337/2b910bbc-8974-11e5-94b4-a32d09491485.JPG)

(oops, semitranslated Swedish, heh)

A little bit more polish and I'll merge, it's pretty neat :)

LunaMoo committed:
Make sure display representation matches all cases and recreate views on window resize(this also fixes touch control layout screen).

LunaMoo commented:
Thanks for finding that out, appears touch control layout screen was affected by that as well so I fixed both.
 Also added some other cases to display representation on auto to match functionality in cases other than common widescreen resolution since that was also missing.

LunaMoo commented:
@hrydgard what do you think about also moving "Stretch to display" checkbox inside display layout editor screen? I could change it to be always accessible, just disable zoom options when stretched instead. I don't think that option is often used either and also fits meaning another -1 to graphics settings screen.

hrydgard commented:
Yeah, that would be great.

LunaMoo committed:
Integrate stretching options into display layout editor.

LunaMoo commented:
Instead of separate option, I renamed zoom settings to more general "Options"(and reused translation for it from cheat menu) and included both stretching options there. They can't be enabled with other options anyway, so might as well be one.
 Also when stretching is used it'll replace the screen representation to reflect that.

unknownbrackets commented on Windows/MainWindow.cpp:
```diff
@@ -281,6 +283,10 @@ namespace MainWindow
 		}
 
+		if (screenManager) {
+			screenManager->RecreateAllViews();
```
Is this the correct thread to do this from?  Shouldn't this happen via an event?

-[Unknown]

LunaMoo replied:
Well that's a fix for the issue which messes up both this and touch control layout screens since they aren't recreated whenever you resize window as on Henrik's screenshot above. Not really sure how to deal with it differently.

unknownbrackets replied:
Hmm.  Maybe we should recreate views when the resized event comes through to the DisplayLayoutEditor?  I mean via `void DisplayLayoutScreen::sendMessage(const char *message, const char *value) {`.

-[Unknown]

unknownbrackets commented on Windows/ppsspp.rc:
```diff
@@ -442,7 +449,22 @@ BEGIN
+		BEGIN
+			MENUITEM "Full stretch", ID_OPTIONS_STRETCH
+			MENUITEM "Partial stretch", ID_OPTIONS_PARTIAL_STRETCH
+			MENUITEM "Auto scaling", ID_OPTIONS_DISPLAY_AUTO
```
I don't know if a layperson will understand the difference between these three options.

-[Unknown]

LunaMoo replied:
Yeah I think it will be better if I will just make it an option to open the screen as there it just shows what each option does clearing any potential confusion and those settings aren't all that useful to change often anyway.

unknownbrackets commented on Core/Config.cpp:
```diff
@@ -443,9 +435,7 @@ static ConfigSetting graphicsSettings[] = {
-	ConfigSetting("PartialStretch", &g_Config.bPartialStretch, &DefaultPartialStretch, true, true),
-	ConfigSetting("StretchToDisplay", &g_Config.bStretchToDisplay, false, true, true),
-	ConfigSetting("SmallDisplayZoom", &g_Config.iSmallDisplayZoom, 0, true, true),
+	ConfigSetting("SmallDisplayZoom", &g_Config.iSmallDisplayZoom, 2, true, true),
```
Hmm, not sure if I'm really a fan of shoving a bunch of different things (including a numeric scale) into the same option.  Especially changing the meaning of "3" from 3x to 1x in an existing option.

Would it make more sense to have "SmallDisplayZoomType" and "SmallDisplayZoomLevel"?

-[Unknown]

LunaMoo replied:
I might have a weird concept of order and simplicity since to me all of those settings being used in one function affecting same thing felt natural, but I don't mind dividing it. I guess types would be same as now plus manual scaling and show zoom level options separately only when manual scaling is selected.

unknownbrackets commented on Core/Config.cpp:
```diff
@@ -443,9 +435,7 @@ static ConfigSetting graphicsSettings[] = {
 #endif
 
 	// TODO: Replace these settings with a list of options
-	ConfigSetting("PartialStretch", &g_Config.bPartialStretch, &DefaultPartialStretch, true, true),
```
It might be nice to retain this default for Blackberry only; my understanding is that the aspect ratio of the display makes the zoom desirable as a default.  @xsacha?

-[Unknown]

xsacha replied:
That's right but only for the keyboard Blackberry 10 devices (there's only 5 of them): Q5, Q10, DAC, Classic, Passport

Inside a 1:1 aspect ratio looks bad without some zoom.

LunaMoo replied:
Yeah, I think I just forgot about this since the blackberry default was set at different place, will change it along the rest after I get some sleep ~ busy part of the year @_@.

LunaMoo committed:
Some improvements, comments and polishing.

LunaMoo commented:
Changes:
 Re-added blackberry partial stretch check that was setting default, but since it really only checks for resolution proportions, I made it for all devices.
 Divided zoom type from level also limited all display representation scaling(*/ 8) to display layout screen and added an extra slider to set zoom level accurately aside to scaling.
 Got rid of the lenghty and potentially confusing submenu and replaced it with direct link to the screen.

Pretty much that's how it looks now:
![displaylayout](https://cloud.githubusercontent.com/assets/5485237/11924396/aaf2fd5c-a7b0-11e5-8517-ceccfa42317c.jpg)
Extra options are hidden when stretching or automatic scaling is used, also rotation is grayed out and visual representation looses potential rotation when non-buffered rendering is being used.

unknownbrackets commented:
I use a 4:3 screen and would not want partial stretch on my desktop.  I would prefer the black bars.  I'm not sure making it the default for everyone makes sense, it makes more sense for small screens.

-[Unknown]

LunaMoo committed:
Set default partial stretch only for blackberry devices without ws.

LunaMoo commented:
Ok then, changed it your way. Althrough imo the same way someone could not like it on smaller device as well since it's a matter of proportions not size really, just like emulating older consoles on psp, small wide screen, but it's not like everyone stretched old games just to get a bit less black bars.
 Either way it's just defaults anyway so anyone wanting to change it will be able.

unknownbrackets commented:
Hey, it's been a while, I think some change has conflicted.  Pretty sure it's my fault - I changed this line `if (g_Config.iGPUBackend == GPU_BACKEND_OPENGL) { offsetY = offsetY * -1.0f; }` to fix a backend thing.

Could you rebase?  I haven't looked carefully at all the UI code, but it looks like this is ready to go, right?

-[Unknown]

LunaMoo commented:
Sorry kind of forgot about it, but yeah should be ready / rebased.

unknownbrackets commented on Qt/mainwindow.cpp:
```diff
@@ -539,8 +543,10 @@ void MainWindow::createMenus()
-		->addEventChecked(&g_Config.bStretchToDisplay);
+	MenuTree* displayLayoutMenu = new MenuTree(this, videoMenu, QT_TR_NOOP("&Display Layout Options"));
+	displayLayoutGroup = new MenuActionGroup(this, displayLayoutMenu, SLOT(displayLayoutGroup_triggered(QAction *)),
+		QStringList() << "Stretched" << "Partialy stretched" << "Auto Scaling" << "Manual Scaling",
```
*Partially.... does this get translated?

-[Unknown]

unknownbrackets commented on UI/DisplayLayoutScreen.cpp:
```diff
@@ -192,37 +191,87 @@ void DisplayLayoutScreen::CreateViews() {
-	
-
+	static const char *zoomLevels[] = { "Stretching", "Partial Stretch", "Auto Scaling", "Manual Scaling" };
+	zoom_ = new PopupMultiChoice(&g_Config.iSmallDisplayZoomType, cw->T("Options"), zoomLevels, 0, ARRAY_SIZE(zoomLevels), gr->GetName(), screenManager(), new AnchorLayoutParams(400, WRAP_CONTENT, local_dp_xres / 2 - 200, NONE, NONE, 10));
```
Hmm.  I'll defer to anyone doing translations, but it seems unsual to use `Options` from another section.

Maybe "Cheating Options" and "Stretching/Scaling Options" might be two separate words in some non-English language, like "High [quality]" and "Tall [display]" are different words in English.

-[Unknown]

unknownbrackets commented on UI/DisplayLayoutScreen.cpp:
```diff
@@ -192,37 +191,87 @@ void DisplayLayoutScreen::CreateViews() {
+	rotation_->SetEnabledPtr(&displayRotEnable_);
+	displayRotEnable_ = (g_Config.iRenderingMode != FB_NON_BUFFERED_MODE);
+	bool bRotated = false;
+	if (g_Config.iRenderingMode != FB_NON_BUFFERED_MODE && (g_Config.iInternalScreenRotation == ROTATION_LOCKED_VERTICAL || g_Config.iInternalScreenRotation == ROTATION_LOCKED_VERTICAL180)) {	bRotated = true; }
```
So long.  Would prefer `displayRotEnable_ && ...` as well as the body of the if on its own line.

-[Unknown]

unknownbrackets commented on UI/DisplayLayoutScreen.cpp:
```diff
@@ -192,37 +191,87 @@ void DisplayLayoutScreen::CreateViews() {
+			Choice *center = new Choice(di->T("Center"), "", false, new AnchorLayoutParams(leftColumnWidth, WRAP_CONTENT, 10, NONE, NONE, 74));
+			center->OnClick.Handle(this, &DisplayLayoutScreen::OnCenter);
+			root_->Add(center);
+			PopupSliderChoiceFloat *zoomlvl_ = new PopupSliderChoiceFloat(&g_Config.fSmallDisplayZoomLevel, 1.0f, 10.0f, di->T("Zoom"), 1.0f, screenManager(), "* PSP res", new AnchorLayoutParams(leftColumnWidth, WRAP_CONTENT, 10, NONE, NONE, 134));
```
Does `"* PSP res"` need a `->T()`?  Not sure if it's translated in `PopupSliderChoiceFloat` or not...

-[Unknown]

unknownbrackets commented on Windows/ppsspp.rc:
```diff
@@ -444,7 +451,7 @@ BEGIN
         MENUITEM "Language...",                             ID_OPTIONS_LANGUAGE
         MENUITEM SEPARATOR
-        MENUITEM "Stretch to Display",                      ID_OPTIONS_STRETCHDISPLAY
+		MENUITEM "Display Layout Editor"                    ID_OPTIONS_DISPLAY_LAYOUT
```
Doesn't really matter, but this file is spaces-land for some msvc-editor reason.

-[Unknown]

unknownbrackets commented on Windows/resource.h:
```diff
@@ -319,6 +318,11 @@
+#define ID_EMULATION_ROTATION_V          40157
+#define ID_EMULATION_ROTATION_H_R        40158
+#define ID_EMULATION_ROTATION_V_R        40159
+#define ID_OPTIONS_DISPLAY_LAYOUT        40160
```
Also not a big thing, but if you're visiting this code, maybe also change `_APS_NEXT_COMMAND_VALUE` to `40161` to keep the visual editor happy.

-[Unknown]

LunaMoo committed:
Tabs can't live in space and some other corrections.

LunaMoo commented:
Fixed all of that not counting Qt, no clue what to do with it, I recall I initially just fixed the build trying to maintain existing functionality, but I had to do it blindly, I don't even know how Qt menu looks(well I imagine it is a drop down menu, but that's it), looking at other options as I based that code on existing stuff I saw there it doesn't seem to have translations.(?)
 Also would be best to just link it to display layout editor menu like on windows, but I don't know if it's even possible for it.

unknownbrackets commented:
> //Edit: oh and the land of spaces had suprising amount of aliens;p.

Hahahaha... sorry.  Thanks.

In the interface, "Stretching" and "Auto Scaling" look different, and "Stretching" reacts to clicks but doesn't do anything.  Is that intentional?

Also, I think it's an improvement as is, but if you click not in the center, and then move, the frame jumps to center on your cursor.  Usually I prefer to save the initial click offset somewhere, and apply that every time so that the drag is fluid.

It'd be cool if you could focus that particular view, and press X, and then use the arrow keys to move it, and then press X again or O to release.

Anyway, those don't seem necessary.  LGTM, what do you think @hrydgard?

-[Unknown]

LunaMoo committed:
Smooth out touch moving and resize and other small changes.

LunaMoo commented:
Don't know about controls, but the moving part was easy and seems nice change, thanks for that idea.

  Also minor changes - displayed zoom value should now update instantly when resizing and disabled that click you catched as that "choice" is there just a visualization.

unknownbrackets commented on UI/DisplayLayoutScreen.cpp:
```diff
@@ -68,50 +71,54 @@ bool DisplayLayoutScreen::touch(const TouchInput &touch) {
 		if (mode == 0) {
 			const Bounds &bounds = picked_->GetBounds();
 
 			int mintouchX = screen_bounds.w / 4;
```
Oh, sorry for the nitpick - `minTouchX` right?

-[Unknown]

LunaMoo replied:
Nah minTouchX is not offset it sets bounds of the movement on the screen, I reused it in maxTouchX because it was same equation, ~ simplified, but that's not offset:).

unknownbrackets replied:
I just meant `touch` vs `Touch`, your change is good I'm just being the camel case police, sorry.

-[Unknown]

LunaMoo replied:
Aaahh I see now, I looked at the changed code instead of the line you mentioned ~ attention colors heh;p. Will rename that.

unknownbrackets commented:
Nice.

-[Unknown]

LunaMoo commented:
Pushed it again with a fix for resize I updated incorrectly(could get to 0;p), either way moving and resizing does work much smoother with the offset now.

hrydgard commented:
Alright, let's merge :)

LunaMoo committed:
Rename variable to match others.
